test/e2e/apps/deployment.go

```diff
@@ -1705,3 +1709,146 @@ func testDeploymentSubresources(ctx context.Context, f *framework.Framework) {
 	framework.ExpectNoError(err, "Failed to get deployment resource: %v", err)
 	framework.ExpectEqual(*(deployment.Spec.Replicas), int32(4), "deployment should have 4 replicas")
 }
+
+func testPodTemplateHashCompatible(ctx context.Context, f *framework.Framework) {
+	ns := f.Namespace.Name
+	c := f.ClientSet
+
+	container := v1.Container{
+		Name:  "my-container",
+		Image: imageutils.GetPauseImageName(),
+	}
+
+	labelSelector := metav1.FormatLabelSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{"test-name": "pod-template-hash-compatible"},
+	})
+
+	checkPodTemplateHash := func(ctx context.Context, podNum int, hasOldLabel, hasNewLabel bool) error {
+		pods, err := c.CoreV1().Pods(ns).List(context.Background(), metav1.ListOptions{
```

Comment:
maybe we could also check the replica set here, that it has the correct labels and selector?